dot/sync/fullsync.go

```diff
@@ -0,0 +1,599 @@
+			req:          request,
+			responseData: response.BlockData,
+		})
+	}
+
+	return repChanges, peersToBlock, validRes
+}
+
+// sortFragmentsOfChain will organise the fragments
+// in a way we can import the older blocks first also guaranting that
+// forks can be imported by organising them to be after the main chain
+//
+// e.g: consider the following fragment of chains
+// [ {17} {1, 2, 3, 4, 5} {6, 7, 8, 9, 10} {8} {11, 12, 13, 14, 15, 16} ]
+//
+// note that we have fragments with single blocks, fragments with fork (in case of 8)
+// after sorting these fragments we end up with:
+// [ {1, 2, 3, 4, 5}  {6, 7, 8, 9, 10}  {8}  {11, 12, 13, 14, 15, 16}  {17} ]
+func sortFragmentsOfChain(fragments [][]*types.BlockData) [][]*types.BlockData {
+	if len(fragments) == 0 {
+		return nil
+	}
+
+	slices.SortFunc(fragments, func(a, b []*types.BlockData) int {
+		if a[0].Header.Number < b[0].Header.Number {
+			return -1
+		}
+		if a[0].Header.Number == b[0].Header.Number {
+			return 0
+		}
+		return 1
+	})
+
+	return fragments
+}
+
+// mergeFragmentsOfChain merges a sorted slice of fragments that forms a valid
+// chain sequente which is the previous is the direct parent of the next block,
+// and keep untouch fragments that does not forms such sequence,
+// take as an example the following sorted slice.
+// [ {1, 2, 3, 4, 5}  {6, 7, 8, 9, 10}  {8}  {11, 12, 13, 14, 15, 16}  {17} ]
+// merge will transform it in the following slice:
+// [ {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17} {8} ]
+func mergeFragmentsOfChain(fragments [][]*types.BlockData) [][]*types.BlockData {
+	if len(fragments) == 0 {
+		return nil
+	}
+
+	mergedFragments := [][]*types.BlockData{fragments[0]}
+	for i := 1; i < len(fragments); i++ {
+		lastMerged := mergedFragments[len(mergedFragments)-1]
+		current := fragments[i]
+
+		if formsSequence(lastMerged[len(lastMerged)-1], current[0]) {
+			mergedFragments[len(mergedFragments)-1] = append(lastMerged, current...)
+		} else {
+			mergedFragments = append(mergedFragments, current)
+		}
+	}
+
+	return mergedFragments
+}
+
+// validBlocksUnderFragment ignore all blocks prior to the given last finalized number
+func validBlocksUnderFragment(highestFinalizedNumber uint, fragmentBlocks []*types.BlockData) []*types.BlockData {
+	startFragmentFrom := -1
+	for idx, block := range fragmentBlocks {
+		if block.Header.Number > highestFinalizedNumber {
+			startFragmentFrom = idx
+			break
+		}
+	}
+
+	if startFragmentFrom < 0 {
+		return nil
+	}
+
+	return fragmentBlocks[startFragmentFrom:]
+}
+
+// formsSequence given two fragments of blocks, check if they forms a sequence
+// by comparing the latest block from the prev fragment with the
+// first block of the next fragment
+func formsSequence(prev, next *types.BlockData) bool {
+	incrementOne := (prev.Header.Number + 1) == next.Header.Number
+	isParent := prev.Hash == next.Header.ParentHash
+
+	return incrementOne && isParent
+}
```

Comment:
```suggestion
func isParent(prev, next *types.BlockData) bool {
	incrementOne := (prev.Header.Number + 1) == next.Header.Number
	isParent := prev.Hash == next.Header.ParentHash

	return incrementOne && isParent
}
```

Could also be a method on `BlockData`.